From b51461b88a0fb4ab60e21fcf7f85503e0a7aade0 Mon Sep 17 00:00:00 2001
From: Antonio de Angelis <Antonio.deAngelis@arm.com>
Date: Thu, 18 Sep 2025 13:02:36 +0100
Subject: [PATCH 2/4] CC3XX: DCU: Check dcu_en against the
 permanent_disable_mask

Regardless of the lifecycle state, there is a permanent disable
mask register against which the required DCU_EN need to be checked.

Upstream-Status: Backport [ab8edf16290fc13aa2eb5f5149235613c4f7c9a0]
Signed-off-by: Antonio de Angelis <antonio.deangelis@arm.com>
Change-Id: I2b4435d6ae7ebb8238987be06ac0c3b40b6dc991

---
 .../cc3xx/low_level_driver/src/cc3xx_dcu.c    | 34 ++++++++++++++++++-
 1 file changed, 33 insertions(+), 1 deletion(-)

diff --git a/platform/ext/target/arm/drivers/cc3xx/low_level_driver/src/cc3xx_dcu.c b/platform/ext/target/arm/drivers/cc3xx/low_level_driver/src/cc3xx_dcu.c
index ce9b1afc4a..089589f278 100644
--- a/platform/ext/target/arm/drivers/cc3xx/low_level_driver/src/cc3xx_dcu.c
+++ b/platform/ext/target/arm/drivers/cc3xx/low_level_driver/src/cc3xx_dcu.c
@@ -39,6 +39,32 @@ static cc3xx_err_t check_dcu_restriction_mask(const uint32_t *val)
     return CC3XX_ERR_SUCCESS;
 }
 
+/**
+ * @brief Check that the requested permissions are in accordance with the
+ *        permanent disable mask. A 1 in the mask means disabled
+ *
+ * @param[in] val Sets of permissions, i.e. host_dcu_en to check as an array of 4 words
+ * @return cc3xx_err_t CC3XX_ERR_SUCCESS or CC3XX_ERR_DCU_MASK_MISMATCH
+ */
+static cc3xx_err_t check_dcu_permanent_disable_mask(const uint32_t *val)
+{
+    size_t idx;
+
+    CC3XX_INFO("permanent_disable_mask: 0x%08x_%08x_%08x_%08x\r\n",
+               P_CC3XX->ao.ao_permanent_disable_mask[0],
+               P_CC3XX->ao.ao_permanent_disable_mask[1],
+               P_CC3XX->ao.ao_permanent_disable_mask[2],
+               P_CC3XX->ao.ao_permanent_disable_mask[3]);
+
+    for (idx = 0; idx < sizeof(P_CC3XX->ao.ao_permanent_disable_mask) / sizeof(uint32_t); idx++) {
+        if (val[idx] & P_CC3XX->ao.ao_permanent_disable_mask[idx]) {
+            return CC3XX_ERR_DCU_MASK_MISMATCH;
+        }
+    }
+
+    return CC3XX_ERR_SUCCESS;
+}
+
 /**
  * @brief Check that the requested permissions are in accordance with the
  *        current status of the DCU locks
@@ -150,7 +176,13 @@ cc3xx_err_t cc3xx_dcu_set_enabled(const uint8_t *permissions_mask, size_t len)
                dcu_en_requested[2],
                dcu_en_requested[3]);
 
-    /* Check the restriction mask for the dcu_en*/
+    /* Check the permanent disable mask for the dcu_en */
+    err = check_dcu_permanent_disable_mask(dcu_en_requested);
+    if (err != CC3XX_ERR_SUCCESS) {
+        return err;
+    }
+
+    /* Check the ICV restriction mask for the dcu_en */
     err = check_dcu_restriction_mask(dcu_en_requested);
     if (err != CC3XX_ERR_SUCCESS) {
         return err;
-- 
2.43.0

